File: synara/src/app/features/matrix-dto/widget.ts
/**
 * Widget / Element Call session DTO.
 */

import type { RoomId, WidgetId } from './ids';
import {
  hasForbiddenWireFields,
  isObject,
  optString,
  reqBoolean,
  reqString,
} from './parseUtil';

export const WIDGET_KINDS = ['element_call', 'custom'] as const;
export type WidgetKind = (typeof WIDGET_KINDS)[number];
const KIND_SET = new Set<string>(WIDGET_KINDS);

export const WIDGET_SESSION_STATES = [
  'idle',
  'creating',
  'active',
  'ending',
  'failed',
] as const;
export type WidgetSessionState = (typeof WIDGET_SESSION_STATES)[number];
const STATE_SET = new Set<string>(WIDGET_SESSION_STATES);

export type WidgetSession = {
  widgetId: WidgetId;
  roomId: RoomId;
  kind: WidgetKind;
  state: WidgetSessionState;
  url?: string;
  hasActiveCall: boolean;
};

export function parseWidgetSession(value: unknown): WidgetSession | null {
  if (!isObject(value) || hasForbiddenWireFields(value)) return null;
  const widgetId = reqString(value, 'widgetId');
  const roomId = reqString(value, 'roomId');
  const url = optString(value, 'url');
  const hasActiveCall = reqBoolean(value, 'hasActiveCall');
  if (
    widgetId === null ||
    roomId === null ||
    url === null ||
    hasActiveCall === null ||
    typeof value.kind !== 'string' ||
    !KIND_SET.has(value.kind) ||
    typeof value.state !== 'string' ||
    !STATE_SET.has(value.state)
  ) {
    return null;
  }
  return {
    widgetId,
    roomId,
    kind: value.kind as WidgetKind,
    state: value.state as WidgetSessionState,
    url,
    hasActiveCall,
  };
}

File: src-tauri/src/matrix/dto/widget.rs
//! Widget / Element Call session DTO.
//!
//! Prefer host-side URL construction later; wire DTO may omit secrets/query tokens.

use serde::{Deserialize, Serialize};

use super::ids::{RoomId, WidgetId};

/// Widget product kind.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WidgetKind {
    ElementCall,
    Custom,
}

impl WidgetKind {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::ElementCall => "element_call",
            Self::Custom => "custom",
        }
    }
}

/// Widget / call session lifecycle.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WidgetSessionState {
    Idle,
    Creating,
    Active,
    Ending,
    Failed,
}

impl WidgetSessionState {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Idle => "idle",
            Self::Creating => "creating",
            Self::Active => "active",
            Self::Ending => "ending",
            Self::Failed => "failed",
        }
    }
}

/// Element Call / custom widget session projection.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WidgetSession {
    pub widget_id: WidgetId,
    pub room_id: RoomId,
    pub kind: WidgetKind,
    pub state: WidgetSessionState,
    /// Optional URL — must not embed access tokens or recovery material.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub url: Option<String>,
    pub has_active_call: bool,
}
